### PR TITLE
refactor: move away from page state architecture

### PR DESCRIPTION
**Pull request by mediremi**
Closes https://jira.dhis2.org/browse/DHIS2-11070

The current architecture of this app, where every page component extends a `Page` class and sets its own state through a callback that then updates its props (instead of `setState`), is unintuitive and impacts development.

This PR:
- removes this `Page` abstraction
- adds Cypress tests for each page
- introduces a `usePoll` hook for polling task statuses. This hook is currently only used for the 'data integrity' page, but as the 'resource tables' and 'analytics' pages also implement task polling its use should be expanded in the future.

---

Additional fixes included in this PR:
- [The documentation links for each page have been fixed](https://github.com/dhis2/data-administration-app/blob/796a3b0dac04e7ff71d290d214a2800b75bc2f43/src/components/DocsLink/DocsLink.js#L24)
- Data statistics page: [All object type labels are now visible](https://github.com/dhis2/data-administration-app/blob/796a3b0dac04e7ff71d290d214a2800b75bc2f43/src/pages/data-statistics/DataStatistics.js#L19-L42) (the `Visualization`, `Tracked entity instances` and `Program instances` were hidden due to not having associated `i18n` strings).
- Update schema of lock exceptions to match `2.37`. `periodType` is now an object of shape `{ name: String }` instead of a string.
- [String literals passed to `i18n.t`](https://github.com/dhis2/data-administration-app/blob/796a3b0dac04e7ff71d290d214a2800b75bc2f43/src/i18n-keys.js) (instead of references [e.g. [here](https://github.com/dhis2/data-administration-app/pull/623/files#diff-b63be53b3a6f61bfa6f6cbb3b2486edfbf16b813380ebb66901416ef840f8fd4L33)]), so translation files show more strings
- Uses the app runtime alerts service instead of a snackbar (the snackbar was previously updated while updating the state of the parent `Page` components and so had to be removed)
- Replaces all `material-ui` buttons with `@dhis2/ui` buttons
- Add confirmation modals for deleting lock exceptions to 'lock exception management' page

### CODE REVIEW

File: src/pages/lock-exception/AddLockExceptionForm.js [hunk 1]
```diff
@@ -119,125 +172,113 @@ class AddLockExceptionForm extends Component {
...
+            </div>
+        )
+    }
+
     render() {
         const dataSetItems = this.props.dataSets.map(dataSet => ({
             id: dataSet.id,
             name: dataSet.displayName,
-            periodType: dataSet.periodType,
+            periodType: dataSet.periodType.name,
```
Comment (mediremi) [inline src/pages/lock-exception/AddLockExceptionForm.js:244]: We'll want to revert this change once https://jira.dhis2.org/browse/DHIS2-11198 is resolved.

File: cypress/fixtures/data-sets.json [hunk 2]
```diff
@@ -0,0 +1,134 @@
+{
+    "dataSets": [
+        {
+            "id": "lyLU2wR22tC",
+            "displayName": "ART monthly summary",
+            "periodType": { "name": "Monthly" }
```
Comment (mediremi) [inline cypress/fixtures/data-sets.json:6]: This fixture will need to be updated once https://jira.dhis2.org/browse/DHIS2-11198 is resolved.

File: cypress/integration/analytics-tables/index.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    )
+})
+
+When(`the user clicks on the 'Start export' button`, () => {
+    cy.findByRole('button', { name: 'Start export' }).click()
+})
+
+Then('the tables are exported', () => {
+    cy.wait('@startAnalyticsTablesExportTask')
+    cy.tick(5000)
```
Comment (ismay) [inline cypress/integration/analytics-tables/index.js:48]: I assume that this is because of the `PULL_INTERVAL` constant? I can imagine that either an explanatory comment, or importing and using the PULL_INTERVAL here would make it clearer what's going on. I think I'd probably lean towards the comment.

Comment (ismay) [reply cypress/integration/analytics-tables/index.js:48]: Or, if there'd be a way to not have to deal with timeouts at all, but instead wait for some content to appear (or something like that), that seems even better (to me).

File: cypress/integration/data-integrity/index.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    cy.findByRole('button', { name: 'Run integrity checks' }).click()
+})
+
+Then('a data integrity check is requested', () => {
+    cy.tick(5000)
+    cy.wait('@pollDataIntegrityCheck')
+})
+
+Then('the user is shown the results of the check', () => {
+    // TODO
```
Comment (ismay) [inline cypress/integration/data-integrity/index.js:46]: Is this a todo we should resolve in the current PR? Or after that?

File: src/App.module.css [hunk 5]
```diff
@@ -0,0 +1,26 @@
+.container {
+    display: grid;
+    height: 100%;
```
Comment (ismay) [inline src/App.module.css:3]: I wonder if the 100% heights are necessary here and for the sidebar. Seems to be fine without.

Comment (mediremi) [reply src/App.module.css:3]: It's not needed for most pages, but the 'Analytics Tables' and 'Min-Max Value Generation' pages seem cutoff without it since the sidebar becomes too short.

| Without 100% height | With 100% height |
--- | ---
| image | image |

Comment (ismay) [reply src/App.module.css:3]: Ah ok I didn't notice that 👍

File: cypress/integration/analytics-tables/index.js (hunk 3 above)
Comment (mediremi) [reply cypress/integration/analytics-tables/index.js:48]: That's right - this is because the first poll occurs after `PULL_INTERVAL` milliseconds. The export button does change its text from 'Start export' to 'Exporting...' once the user clicks on it, which we could use instead of waiting for the poll to occur, but I'd like to verify that a network request to the right endpoint occurs so I'll go with the comment.

File: src/pages/data-integrity/DataIntegrity.js [hunk 6]
```diff
@@ -1,172 +1,112 @@
...
+}
+
+const pollQuery = {
+    poll: {
+        resource: 'system/taskSummaries/DATA_INTEGRITY',
+        id: ({ jobId }) => jobId,
+    },
+}
+
+const usePoll = (fn, pollInterval) => {
```
Comment (ismay) [inline src/pages/data-integrity/DataIntegrity.js:22]: Might be nice to extract the hooks so that this file doesn't become too big.

File: src/pages/lock-exception/AddLockExceptionModal.js [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+        } catch (error) {
+            errorAlert.show({ error })
+        }
+        setIsLoading(false)
+    }
+    const isAddActionDisabled =
+        isLoading ||
+        !(selectedOrgUnits.length > 0 && selectedDataSetId && selectedPeriodId)
+
+    // TODO: Need to use material UI Dialog component instead of @dhis2/ui Modal
```
Comment (ismay) [inline src/pages/lock-exception/AddLockExceptionModal.js:52]: I wonder if we could also use the @dhis2/ui Layer component to remedy this issue. It's meant to give control over the z-index, though I'm not sure if it would help here.

File: src/pages/maintenance/Maintenance.js [hunk 8]
```diff
@@ -1,180 +1,126 @@
...
-                            },
-                        })
-                    }
-                })
-                .catch(error => {
-                    if (this.isPageMounted()) {
-                        this.setLoadedPageWithErrorState(error)
-                    }
-                })
+const Maintenance = ({ sectionKey }) => {
```
Comment (ismay) [inline src/pages/maintenance/Maintenance.js:61]: This file is also getting fairly large. I think splitting it up would help with the readability.

File: cypress/integration/analytics-tables/index.js (hunk 3 above)
Comment (ismay) [reply cypress/integration/analytics-tables/index.js:48]: Sounds good 👍

**Review by ismay — COMMENTED** [review]
Looks good, nice work. Integration tests all pass for me. I left a couple of comments on details that stood out to me. Let me know what you think and then it should be good to merge I think!

File: cypress/integration/data-integrity/index.js (hunk 4 above)
Comment (mediremi) [reply cypress/integration/data-integrity/index.js]: Both - I was unsure of the best way to test this and I was hesistant to try as the issue cards are material UI components which we should replace with `@dhis2/ui` components (which would change the way we test this). For now I've added some basic tests that only verify the card titles, but once we migrate away from material UI it'd be nice to also check the content of the cards (shown when they are expanded by the user).

File: src/pages/lock-exception/AddLockExceptionModal.js (hunk 7 above)
Comment (mediremi) [reply src/pages/lock-exception/AddLockExceptionModal.js:52]: The issue with the select inputs from material ui is that they render their dropdowns as popovers, so the z-index of the associated input component has no effect on their position relative to the modal. The best solution imo would be to replace all material ui components used in the data administration app with ones from `@dhis2/ui`, which would mean either reimplementing the organisation selection buttons from `d2-ui` or copy & pasting their code and changing their select field import from `material-ui` to `@dhis2/ui`

**Review by ismay — APPROVED** [review]
Sorry for the late reply, things kept popping up. Thanks for the explanatory comments, all looks good to me 👍!

One thing that might be good as a follow up is to create jira issues for any technical debt you've encountered (or any other things that still need to be addressed). If you have the time, I think that that would be very helpful as you're now probably quite familiar with the codebase.

Another I noticed is that for me these integration tests failed locally (when running cy:run):

Screenshot 2021-06-01 at 17 22 36

Screenshot 2021-06-01 at 17 22 25

Once those pass I think this should be good to merge 👍

**Comment by mediremi** [comment]
> One thing that might be good as a follow up is to create jira issues for any technical debt you've encountered (or any other things that still need to be addressed)

Good idea :+1:  - I've created the epic https://jira.dhis2.org/browse/DHIS2-11246 along with 5 linked Jira issues.

> Another I noticed is that for me these integration tests failed locally (when running cy:run):

I think this might be due to an incorrect `cypress.env.json` config, specifically the `dhis2_datatest_prefix` field.

The following config results in passing tests for me (tested with Chromium and Electron):
```
{
    "dhis2_base_url": "https://debug.dhis2.org/dev",
    "dhis2_username": "admin",
    "dhis2_password": "district",
    "dhis2_datatest_prefix": ""
}
```

While this one results in the same failing tests as you encountered:
```
{
    "dhis2_base_url": "https://debug.dhis2.org/dev",
    "dhis2_username": "admin",
    "dhis2_password": "district",
    "dhis2_datatest_prefix": "not-correct-prefix"
}
```

**Comment by ismay** [comment]
> > One thing that might be good as a follow up is to create jira issues for any technical debt you've encountered (or any other things that still need to be addressed)
> 
> Good idea 👍 - I've created the epic https://jira.dhis2.org/browse/DHIS2-11246 along with 5 linked Jira issues.
> 
> > Another I noticed is that for me these integration tests failed locally (when running cy:run):
> 
> I think this might be due to an incorrect `cypress.env.json` config, specifically the `dhis2_datatest_prefix` field.
> 
> The following config results in passing tests for me (tested with Chromium and Electron):
> 
> ```
> {
>     "dhis2_base_url": "https://debug.dhis2.org/dev",
>     "dhis2_username": "admin",
>     "dhis2_password": "district",
>     "dhis2_datatest_prefix": ""
> }
> ```
> 
> While this one results in the same failing tests as you encountered:
> 
> ```
> {
>     "dhis2_base_url": "https://debug.dhis2.org/dev",
>     "dhis2_username": "admin",
>     "dhis2_password": "district",
>     "dhis2_datatest_prefix": "not-correct-prefix"
> }
> ```

Thanks. I tried again with the added empty datatest prefix (I didn't have that before), but still had a test failure. However, it's not consistent. A second run yielded a pass for the test that was failing previously and failed somewhere else.

I suspect this is due to our backend or our autologin plugin being flaky, as both test failures were login errors. I think it's safe to take another look at this at a later stage, the failures don't seem to be due to any flaws in the current PR I'd say.
